# Exercícios Avaliativos R

* **Matéria:** Introdução a Linguagens Estatísticas
* **Professor:** Leonardo Vilela
* **Aluno:** Leonardo Flores

**CIÊNCIA DE DADOS & BIG DATA - PUC MINAS 2017/1**

## Lista de Exercícios 1

### Exercício 1.1

Crie um vetor chamado vec contendo os elementos {26, NA, 72, 45, 12, 16, NA, 88} e obtenha o que se pede:

**a)** Obtenha um novo vetor (vec1) contendo o primeiro e o último elementos do vetor vec;

**b)** Insira entre os elementos 72 e 45 do vetor vec os elementos {24, 13, 42}. Dica: a função append()
pode ser útil;

**c)** Obtenha o produto dos elementos do vetor vec;

**d)** Obtenha os valores que não sejam NA;

**e)** Obtenha os valores maiores que a média dos elementos do vetor vec.

**f)** Obtenha um vetor contendo os valores que imediatamente antecedem e sucedem a cada NA. Dica: a função which() pode ser útil;

**g)** Dadososvetoresa <- c(14, 27, 45, 72, 19, -12)eb <- c(6, 9, 8, NA, 7, 5),calculea + b + 5;

**h)** Crie um vetor num_pares contendo os números inteiros pares menores que 100;

**i)** Exclua deste vetor os valores maiores que 60 e menores que 80;

**j)** Encontre a variância dos valores remanescentes.


In [1]:
# Definindo vetor antes de tudo
vec <- c(26, NA, 72, 45, 12, 16, NA, 88)
vec

[1] 26 NA 72 45 12 16 NA 88

In [2]:
# a) Obtenha um novo vetor (vec1) contendo o primeiro e o último elementos do vetor vec;
tamanho <- length(vec)
vec1 <- c(vec[1], vec[tamanho])
vec1

[1] 26 88

In [3]:
# b) Insira entre os elementos 72 e 45 do vetor vec os elementos {24, 13, 42}. Dica: a função append()
# pode ser útil;

append(vec, c(24, 13, 42), after=3)

[1] 26 NA 72 24 13 42 45 12 16 NA 88

In [4]:
# c) Obtenha o produto dos elementos do vetor vec;

prod(vec, na.rm = TRUE)

[1] 1423319040

In [5]:
# d) Obtenha os valores que não sejam NA;

vec[!is.na(vec)]

[1] 26 72 45 12 16 88

In [6]:
# e) Obtenha os valores maiores que a média dos elementos do vetor vec.
media <- mean(!is.na(vec))
vec[vec > media ]

[1] 26 NA 72 45 12 16 NA 88

In [7]:
# f) Obtenha um vetor contendo os valores que imediatamente antecedem e sucedem a cada NA. Dica: a função which() pode ser útil;
na_indexes <- which(is.na(vec))

first_NA = na_indexes[1]
second_NA = na_indexes[2]


# antecessores do primeiro NA
vec[0: (first_NA - 1)]

# sucessores do primeiro NA
vec[(first_NA + 1): length(vec)]

# antecessores do segundo NA
vec[(second_NA - 1):1]

# sucessores do segundo NA
vec[(second_NA + 1): length(vec)]

[1] 26

[1] 72 45 12 16 NA 88

[1] 16 12 45 72 NA 26

[1] 88

In [8]:
# g) Dadososvetoresa <- c(14, 27, 45, 72, 19, -12)eb <- c(6, 9, 8, NA, 7, 5),calculea + b + 5;

a <- c(14, 27, 45, 72, 19, -12)
b <- c(6, 9, 8, NA, 7, 5)

a + b + 5

[1] 25 41 58 NA 31 -2

In [9]:
# h) Crie um vetor num_pares contendo os números inteiros pares menores que 100;
num_pares <- seq(0, 99, 2)
num_pares

[1]  0  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48
[26] 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94 96 98

In [10]:
# i) Exclua deste vetor os valores maiores que 60 e menores que 80;
x <- num_pares[num_pares > 60 & num_pares < 80]
remanescentes <- num_pares[! num_pares %in% x]
remanescentes

[1]  0  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48
[26] 50 52 54 56 58 60 80 82 84 86 88 90 92 94 96 98

In [11]:
# j) Encontre a variância dos valores remanescentes.

var(remanescentes)

[1] 914.2439

### Exercício 1.2

O volume de um cone de raio R e altura H é dado por 1 × π × R2 × H. Tendo obtido o volume e a 3
altura de seis cones, apresentados a seguir, calcule o volume de cada um deles.

R = {2.27, 1.98, 1.69, 1.88, 1.64, 2.14}

H = {8.28, 8.04, 9.06, 8.70, 7.58, 8.34}

In [12]:
R <- c(2.27, 1.98, 1.69, 1.88, 1.64, 2.14)
H <- c(8.28, 8.04, 9.06, 8.70, 7.58, 8.34)

R ** 2

area = (1/3 * pi * R**2 * H)
area

[1] 5.1529 3.9204 2.8561 3.5344 2.6896 4.5796

[1] 44.67974 33.00768 27.09756 32.20057 21.34939 39.99652

### Exercício 1.3

Calcule o que se pede:

a) 3+9/5

b) 564 + 29

c) 76

d) ln12345

e) log2 12345

f) log10 12345

g)  6  - combinatório de seis, dois a dois (Dica: existe uma função que calcula isso). 

h) |18−50|

i) sin(π),cos(π),tan(π)

j) Arredonde o valor de π para 4 casas decimais;

In [13]:
# Letra A
3 + 9/5

[1] 4.8

In [14]:
# Letra B
56 ** 4 + 29/5

[1] 9834502

In [15]:
# Letra C
sqrt(76)

[1] 8.717798

In [16]:
# Letra D
log(12345)

[1] 9.421006

In [17]:
# Letra E
log2(12345)

[1] 13.59164

In [18]:
# Letra F
log10(12345)

[1] 4.091491

In [19]:
# Letra G
combn(6, 2)

[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,] 1    1    1    1    1    2    2    2    2    3     3     3     4     4    
[2,] 2    3    4    5    6    3    4    5    6    4     5     6     5     6    
     [,15]
[1,] 5    
[2,] 6

In [20]:
# Letra H
calc <- 18 - 50
abs(calc)

[1] 32

In [21]:
# Letra I

sin(pi)
cos(pi)
tan(pi)

[1] 1.224647e-16

[1] -1

[1] -1.224647e-16

In [22]:
# Letra J

round(pi, 4)

[1] 3.1416

### Exercício 1.4

Dada a matriz A definida a seguir, obtenha o que se pede:


A=


|35 72 45| 

|20 9 17 |

|13 16 78|

a) obtenha os elementos da diagonal principal. Dica: existe uma função que faz isso. . . ; 

b) obtenha a soma dos elementos da matriz;

c) obtenha a soma dos elementos da primeira linha e da segunda coluna (A11 +A12 +A13 +A12 +A22 +A32) 

d) obtenha a soma da primeira coluna com com a terceira linha (A.1 + A3.)

e) faça a trasposição da matriz;

f) crie duas novas colunas com os valores: coluna 4:{27, 48, 23}, coluna 5:{12, 3, 9};

g) substitua a terceira coluna de A pela soma da segunda e terceira colunas;

In [23]:
A = matrix(c(35, 20, 13, 72, 9, 16, 45, 17, 78), nrow=3, ncol=3)
A

[,1] [,2] [,3]
[1,] 35   72   45  
[2,] 20    9   17  
[3,] 13   16   78

In [24]:
# a) obtenha os elementos da diagonal principal. Dica: existe uma função que faz isso. . . ;

diag(A)

[1] 35  9 78

In [25]:
# b) obtenha a soma dos elementos da matriz;

sum(A)

[1] 305

In [26]:
# c) obtenha a soma dos elementos da primeira linha e da segunda coluna (A11 +A12 +A13 +A12 +A22 +A32)

first_row = A[1,]
second_column = A[, 2]

sum(first_row + second_column)

[1] 249

In [27]:
# d) obtenha a soma da primeira coluna com com a terceira linha (A.1 + A3.)

first_column = A[, 1]
third_row = A[3,]

sum(first_column + third_row)

[1] 175

In [28]:
# e) faça a transposição da matriz.

t(A)

[,1] [,2] [,3]
[1,] 35   20   13  
[2,] 72    9   16  
[3,] 45   17   78

In [29]:
# f) crie duas novas colunas com os valores: coluna 4:{27, 48, 23}, coluna 5:{12, 3, 9};

A <- cbind(A, c(27, 48, 23)) 
A <- cbind(A, c(12, 3, 9))
A

[,1] [,2] [,3] [,4] [,5]
[1,] 35   72   45   27   12  
[2,] 20    9   17   48    3  
[3,] 13   16   78   23    9

In [30]:
# g) substitua a terceira coluna de A pela soma da segunda e terceira colunas;

A[, 3] = A[, 2] + A[, 3]
A

[,1] [,2] [,3] [,4] [,5]
[1,] 35   72   117  27   12  
[2,] 20    9    26  48    3  
[3,] 13   16    94  23    9

### Exercício 1.5

a) Obtenha as linhas para as quais existam NA. Dica: a função complete.cases() pode ser útil; 

b) Obtenha a soma dos valores da coluna idade;

c) Selecione a linha contendo o nome da pessoa de menor idade;

d) Obtenha as linas para as quais não existam NA;

e) Calcule as dimensões de df['idade'] e df[['idade']]. Use a função dim(); 

f) Obtenha os nomes das variáveis do data frame df;

g) Adicione ao data frame os seguintes valores {Felipe, 36}, {Carla, 49} e {Rosane, 18}; 

h) Obtenha os números das linhas para as quais existam NA;

In [31]:
df <- data.frame(
           nome = c('Marcos', 'João', 'José', NA, 'Maria', 'Helena', 'Jordana'),
           idade = c(44, 26, NA, 29, 41, 55, 72),
            stringsAsFactors=FALSE)
df

nome    idade
1 Marcos  44   
2 João    26   
3 José    NA   
4 NA      29   
5 Maria   41   
6 Helena  55   
7 Jordana 72

In [32]:
# a) Obtenha as linhas para as quais existam NA. Dica: a função complete.cases() pode ser útil;

df[!complete.cases(df),]

nome idade
3 José NA   
4 NA   29

In [33]:
# b) Obtenha a soma dos valores da coluna idade;

sum(df[, "idade"], na.rm = T)

[1] 267

In [34]:
# c) Selecione a linha contendo o nome da pessoa de menor idade;

min_idade = min(df[, "idade"], na.rm = T)

df[which(df$idade == min_idade),]

nome idade
2 João 26

In [35]:
# d) Obtenha as linas para as quais não existam NA;

df[complete.cases(df),]

nome    idade
1 Marcos  44   
2 João    26   
5 Maria   41   
6 Helena  55   
7 Jordana 72

In [36]:
# e) Calcule as dimensões de df['idade'] e df[['idade']]. Use a função dim();

dim(df['idade'])
dim(df[['idade']])

[1] 7 1

NULL

In [37]:
# f) Obtenha os nomes das varáveis do dataframe df

colnames(df)

[1] "nome"  "idade"

In [38]:
# g) Adicione ao data frame os seguintes valores {Felipe, 36}, {Carla, 49} e {Rosane, 18};

# Data 1
new_row1 <- c("Felipe", 36)
new_row2 <- c("Carla", 49)
new_row3 <- c("Rosane", 18)

df <- rbind(df, new_row1)
df <- rbind(df, new_row2)
df <- rbind(df, new_row3)
df

nome    idade
1  Marcos  44   
2  João    26   
3  José    NA   
4  NA      29   
5  Maria   41   
6  Helena  55   
7  Jordana 72   
8  Felipe  36   
9  Carla   49   
10 Rosane  18

In [39]:
# h) Obtenha o número de linhas para as quais existam NA.

nrow(df[!complete.cases(df),])

[1] 2

### Exercício 1.6 
Considere que você possua dados sobre o peso (em kg) e altura (em cm) de um conjunto de pessoas, conforme mostrado na tabela abaixo.

Calcule o índice de massa corporal das 4 pessoas e identifique qual delas está com o IMC maior que 25.

In [40]:
people <- matrix(c(
    "Pedro", 180, 87, 
    "José", 165 , 58,
    "Maria", 160, 65, 
    "Sandra", 193, 100
    ), ncol=3, byrow=TRUE)


colnames(people) <- c("Nome", "Altura", "Peso")

people <- as.data.frame(people, stringAsFactor=F)
 
people_heights <- as.numeric(as.character(people$Altura)) / 100 # cm
people_weights <- as.numeric(as.character(people$Peso))

people$IMC <- people_weights / people_heights ** 2

people[which(people$IMC > 25),]

Nome   Altura Peso IMC     
1 Pedro  180    87   26.85185
3 Maria  160    65   25.39062
4 Sandra 193    100  26.84636

### Exercício 1.7

Experimente o comando which.min(c(3,5,1,7,8)). Explique o valor retornado pela função.

**RESPOSTA:** Está sendo retornado o menor valor do VETOR! O **which** procura por um valor e neste caso, o método **min** pega o valor mínimo.

In [41]:
which.min(c(3,5,1,7,8))

[1] 3

### Exercício 1.8

Considere a lista x.lis definida da seguinte forma: x.lis <- list(a=1:10, b=letters[1:3], c=matrix(1:10,ncol=2)).

a) Obtenha a soma acumulada dos elementos da primeira componente; 

b) Obtenha a segunda componente;

c) Obtenha os nomes das componentes da lista;

d) Adicione o vetor vec criado anteriormente à lista;

In [42]:
x.lis <- list(a=1:10, b=letters[1:3], c=matrix(1:10,ncol=2))
x.lis$a

[1]  1  2  3  4  5  6  7  8  9 10

In [43]:
# a) Obtenha a soma acumulada dos elementos da primeira componente;
sum(x.lis$a)

[1] 55

In [44]:
# b) Obtenha a segunda componente;
x.lis$b

[1] "a" "b" "c"

In [45]:
# c) Obtenha os nomes das componentes da lista;
names(x.lis)

[1] "a" "b" "c"

In [46]:
# d) Adicione o vetor vec criado anteriormente à lista;
x.lis$vec <- c(26, NA, 72, 45, 12, 16, NA, 88)
x.lis

$a
 [1]  1  2  3  4  5  6  7  8  9 10

$b
[1] "a" "b" "c"

$c
     [,1] [,2]
[1,]    1    6
[2,]    2    7
[3,]    3    8
[4,]    4    9
[5,]    5   10

$vec
[1] 26 NA 72 45 12 16 NA 88

## Lista de Exercícios 2

### Antes de tudo, configurar o working directory que contém os arquivos

In [47]:
wd <- "/Users/leonardoflores/Dropbox/PUC/ILE/Dados_Curso_R"
setwd(wd)

### Exercício 2.1

No portal da transparência do governo federal na seção denominada Download de dados é possível baixar diversos conjuntos de dados disponíveis no portal. Dentre estes dados consta o CADASTRO NA- CIONAL DE EMPRESAS INIDÔNEAS E SUSPENSAS (CEIS), que é um arquivo contendo a relação das empreas que foram declaradas inidôneas ou foram suspensas de contratar com a administração pública.

Importe o conjunto de dados contido no arquivo 20160306_CEIS.zip e informe quantos registros tem a base de dados.

In [48]:
library(readr)
filename <- '20160306_CEIS.csv'


CEISData <- read.csv(file=filename, sep="\t", fileEncoding="ISO-8859-1")

n_rows <- NROW(CEISData)

sprintf("O número de registros presentes neste conjunto de dados do CEIS é equivalente à %i registros!", n_rows)


Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :
“line 2 appears to contain embedded nulls”Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :
“line 3 appears to contain embedded nulls”Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :
“line 4 appears to contain embedded nulls”Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :
“line 5 appears to contain embedded nulls”Warning message in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, :
“embedded nul(s) found in input”

[1] "O número de registros presentes neste conjunto de dados do CEIS é equivalente à 8853 registros!"

### Exercício 2.2

O Tribunal de Contas da União vem colhendo informações junto a seus jurisdicionados visando obter um perfil da Governança de TI dos mesmos. Os dados relativos ao ciclo de 2014 estão disponíveis no endereço eletrônico http://dados.gov.br/dataset/dados-coletados-no-perfil-de-governanca-de-ti.

Importe o conjunto de dados DadosColetados_PerfilGovTI 2014.xlsx e diga quantos registros tem a base de dados.

In [49]:
library("readxl")

filename <- "DadosColetados_PerfilGovTI\ 2014.xlsx"

PGovTiData <- read_excel(filename, sheet = 1, skip=2)

n_rows <- NROW(PGovTiData)

sprintf("O número de registros presentes neste conjunto de dados do PerfilGovTi é equivalente à %i registros!", n_rows)


[1] "O número de registros presentes neste conjunto de dados do PerfilGovTi é equivalente à 372 registros!"

### Exercício 2.3

Utilizando o pacote sqldf e o conjunto de dados ESCOLAS.CSV importe os registros relativos às escolas do município do Rio de Janeiro.
Para fazer esta seleção as variáveis FK_COD_MUNICIPIO e ID_DEPENDENCIA_ADM devem ser, respectivamente, iguais á 3304557 e 3.


In [50]:
library(sqldf)

filename <- 'ESCOLAS.csv'

escolas <- read.csv(file=filename, sep="|")

sqldf('select * from escolas WHERE FK_COD_MUNICIPIO = 3304557 AND ID_DEPENDENCIA_ADM = 3')


Loading required package: gsubfn
Loading required package: proto
Loading required package: RSQLite
Loading required package: tcltk
Warning message:
“Quoted identifiers should have class SQL, use DBI::SQL() if the caller performs the quoting.”

ANO_CENSO PK_COD_ENTIDADE
1    2014      33066329       
2    2014      33164800       
3    2014      33081760       
4    2014      33087130       
5    2014      33087237       
6    2014      33066515       
7    2014      33130370       
8    2014      33076014       
9    2014      33076022       
10   2014      33076049       
11   2014      33081247       
12   2014      33104824       
13   2014      33087040       
14   2014      33074429       
15   2014      33074437       
16   2014      33070610       
17   2014      33070628       
18   2014      33083347       
19   2014      33083363       
20   2014      33147345       
21   2014      33147353       
22   2014      33075964       
23   2014      33080640       
24   2014      33149283       
25   2014      33078262       
26   2014      33078300       
27   2014      33084114       
28   2014      33083371       
29   2014      33078238       
30   2014      33084610       
⋮    ⋮         ⋮              
1471 2014      33163642       
1472 2014      33085080       
1473 2014      33094551       
1474 2014      33165688       
1475 2014      33163600       
1476 2014      33126577       
1477 2014      33147370       
1478 2014      33068739       
1479 2014      33068771       
1480 2014      33062439       
1481 2014      33168016       
1482 2014      33080097       
1483 2014      33144621       
1484 2014      33122768       
1485 2014      33088055       
1486 2014      33080712       
1487 2014      33065624       
1488 2014      33074313       
1489 2014      33068232       
1490 2014      33144737       
1491 2014      33078270       
1492 2014      33083398       
1493 2014      33078360       
1494 2014      33080887       
1495 2014      33072639       
1496 2014      33082960       
1497 2014      33164134       
1498 2014      33072965       
1499 2014      33069468       
1500 2014      33164142       
     NO_ENTIDADE                                                   
1    0107502 ESCOLA MUNICIPAL MESTRE WALDEMIRO                     
2    POLO DE EDUCACAO PELO TRABALHO ALZIRA ARAUJO                  
3    0918510 CIEP ARMINDO MARCILIO DOUTEL DE ANDRADE               
4    0625029 ESCOLA MUNICIPAL JOSE PEDRO VARELA                    
5    0622008 ESCOLA MUNICIPAL NARBAL FONTES                        
6    0208501 CIEP SAMUEL WAINER                                    
7    0716605 CRECHE MUNICIPAL SEMPRE VIDA NISE DA SILVEIRA         
8    0515052 ESCOLA MUNICIPAL AZEVEDO JUNIOR                       
9    0515034 ESCOLA MUNICIPAL EM 0515034 BARAO DE ITARARE          
10   0515054 ESCOLA MUNICIPAL CARDEAL ARCOVERDE                    
11   0817024 ESCOLA MUNICIPAL WALDIR AZEVEDO FRANCO                
12   1019612 CRECHE MUNICIPAL DANIELA PEREZ                        
13   0625026 ESCOLA MUNICIPAL EUGENIA DUTRA HAMANN                 
14   0514026 ESCOLA MUNICIPAL FRANCISCO SERTORIO PORTINHO          
15   0514023 ESCOLA MUNICIPAL GASPAR VIANNA                        
16   0431026 ESCOLA MUNICIPAL HERBERT MOSES                        
17   0411014 ESCOLA MUNICIPAL JOAO DE DEUS                         
18   0918064 ESCOLA MUNICIPAL RAINHA VITORIA                       
19   0918049 ESCOLA MUNICIPAL RIO GRANDE DO NORTE                  
20   0123601 CRECHE MUNICIPAL JOSE MARINHO DE OLIVEIRA             
21   0328604 CRECHE MUNICIPAL COMUNIDADE DO JACAREZINHO            
22   0515042ESCOLA ESPECIAL MUNICIPAL MAURICIO DE MEDEIROS         
23   0817009 ESCOLA MUNICIPAL JOSE MARIA BELLO                     
24   0515064 ESCOLA MUNICIPAL MESTRE DARCY DO JONGO                
25   0716053 ESCOLA MUNICIPAL PROFESSORA DYLA SYLVIA DE SA         
26   0724019 ESCOLA MUNICIPAL PROFESSOR TEOFILO MOREIRA DA COSTA   
27   1019211 CIEP MAJOR MANOEL GOMES ARCHER                        
28   0918009 ESCOLA MUNICIPAL ROMULO GALLEGOS                      
29   0716038 ESCOLA MUNICIPAL PEDRO AMERICO                        
30   1019014 ESCOLA MUNICIPAL CORONEL BERTH

### Exercício 2.4

No site do Sistema de Inoformações Contábeis e Fiscais(Siconfi) é possível baixar diversos conjuntos de dados contábeis e fiscais de estados e municípios.
O arquivo finbraRREO_MUN_Receitarealizadatotal(ultimos12meses).zip, obtido no site em referência, contém um arquivo .csv com informações relativas ao Anexo 03-Demonstrativo da Receita Corrente Líquida para o primeiro bimestre de 2015. Importe o conteúdo do arquivo. Quantos registros tem o conjunto de dados?


In [51]:
library(sqldf)

filename <- 'finbraRREO.csv'

FinbraRREOData <- read.csv(file=filename, sep=";", skip=5, fileEncoding="ISO-8859-1")

n_rows <- NROW(FinbraRREOData)

sprintf("O número de registros presentes neste conjunto de dados do FinbraRREOData é equivalente à %i registros!", n_rows)


[1] "O número de registros presentes neste conjunto de dados do FinbraRREOData é equivalente à 41133 registros!"

### Exercício 2.5

O arquivo SICONFI_DCA_6561_ANUAL_1.xls contém dados contábeis do município de Angra dos Reis relativos ao exercício de 2014. Utilizando qualquer um dos pacotes para importação de arquivos em excel vistos em sala de aula, importe o conteúdo da planilha DCA-Anexo I-G.

In [52]:
library("readxl")

filename <- "SICONFI_DCA_6561_ANUAL_1.xls"

SICONFIData <- read_excel(filename, skip=17, sheet = "DCA-Anexo I-G")

# Rename first column to 'Despesas por Função'
colnames(SICONFIData)[1] = 'Despesas por Função'

SICONFIData

Despesas por Função                                                                           
1   Despesas (Exceto Intraorçamentárias)                                                          
2   01 - Legislativa                                                                              
3   01.031 - Ação Legislativa                                                                     
4   01.032 - Controle Externo                                                                     
5   01.999 - Demais Subfunções Legislativa                                                        
6   02 - Judiciária                                                                               
7   02.061 - Ação Judiciária                                                                      
8   02.062 - Defesa do Interesse Público no Processo Judiciário                                   
9   02.999 - Demais Subfunções Judiciária                                                         
10  03 - Essencial à Justiça                                                                      
11  03.091 - Defesa da Ordem Jurídica                                                             
12  03.092 - Representação Judicial e Extrajudicial                                               
13  03.999 - Demais Subfunções Essencial à Justiça                                                
14  04 - Administração                                                                            
15  04.121 - Planejamento e Orçamento                                                             
16  04.122 - Administração Geral                                                                  
17  04.123 - Administração Financeira                                                             
18  04.124 - Controle Interno                                                                     
19  04.125 - Normatização e Fiscalização                                                          
20  04.126 - Tecnologia da Informação                                                             
21  04.127 - Ordenamento Territorial                                                              
22  04.128 - Formação de Recursos Humanos                                                         
23  04.129 - Administração de Receitas                                                            
24  04.130 - Administração de Concessões                                                          
25  04.131 - Comunicação Social                                                                   
26  04.999 - Demais Subfunções Administração                                                      
27  05 - Defesa Nacional                                                                          
28  05.151 - Defesa Área                                                                          
29  05.152 - Defesa Naval                                                                         
30  05.153 - Defesa Terrestre                                                                     
⋮   ⋮                                                                                             
147 26.781 - Transporte Aéreo                                                                     
148 26.782 - Transporte Rodoviário                                                                
149 26.783 - Transporte Ferroviário                                                               
150 26.784 - Transporte Hidroviário                                                               
151 26.785 - Transportes Especiais                                                                
152 26.999 - Demais Subfunções Transporte                                                         
153 27 - Desporto e Lazer                                                                         
154 27.811 - Desporto de Rendimento                                                               
155 27.812 - Desporto Comunitário           

### Exercício 2.6

O R dispõe dos conjuntos de dados rock e ToothGrowth. Exporte estes dados para um arquivo do Excel, de forma que cada conjunto de dados fique em uma planilha diferente.

In [54]:
library("xlsx")

write.xlsx(x = ToothGrowth, file = "data.xlsx",
           sheetName = "ToothGrowth", row.names = FALSE)

write.xlsx(x = rock, file = "data.xlsx",
           sheetName = "Rock", row.names = FALSE, append=TRUE)

ERROR: Error in library("xlsx"): there is no package called ‘xlsx’


### Exercício 2.7

Importe os dados contidos no arquivo 14SerieRouboVeiculo2015.xls obtido no site do Instituto de Segurança Pública (ISP). Ignore a coluna e linha de totalizações.

In [55]:
library("readxl")

filename <- "14SerieRouboVeiculo2015.xls"

SerieRouboData <- read_excel(filename, range = "B7:O146")

SerieRouboData

UOp     UNIDADES POLICIAIS                JAN FEV MAR ABR MAI JUN JUL AGO
1   5º BPM  001a. Praça Mauá                    1   1   0  2   1    0   0  2 
2   5º BPM  004a. Praça da República            7   8   5 12   6    2   9  3 
3   5º BPM  005a. Mem de Sá                     5   0   7  1   0    1   3  1 
4   4º BPM  006a. Cidade Nova                   9   4   6  7  14    3   8  2 
5   5º BPM  007a. Santa Tereza                 12   3   5  1   3    2   2  1 
6   2º BPM  009a. Catete                       14   7  12 10  11    4  12  3 
7   2º BPM  010a. Botafogo                      4   4   8 15   4   11   3  3 
8   23º BPM 011a. Rocinha                       0   0   0  0   0    0   0  0 
9   19º BPM 012a. Copacabana                    0   0   0  1   0    0   0  0 
10  19º BPM 013a. Ipanema                       3   0   1  1   0    0   0  0 
11  23º BPM 014a. Leblon                        3   2   2  3   1    0   0  1 
12  23º BPM 015a. Gávea                         1   5   5  2   6    3   0  1 
13  31º BPM 016a. Barra da Tijuca              15  13  12 18  13   11   8  8 
14  4º BPM  017a. São Cristóvão                15  18  22 22  27   25  17 15 
15  4º BPM  018a. Praça da Bandeira            10  15  13 14  21   20  19 16 
16  6º BPM  019a. Tijuca                        7   4  12  7   9    6  13  8 
17  6º BPM  020a. Grajaú                       13  16  15 17   8    4  18 10 
18  22º BPM 021a. Bonsucesso                   34  23  28 29  21   19  20 20 
19  16º BPM 022a. Penha                        33  25  36 43  42   37  41 32 
20  3º BPM  023a. Meier                        38  32  30 20  16   14  21 18 
21  3º BPM  024a. Piedade                      27  24  24 27  29   25  30 27 
22  3º BPM  025a. Engenho Novo                 19  19  33 28  20   15  11 20 
23  3º BPM  026a. Todos os Santos              35  23  26 29  33   24  23 32 
24  41º BPM 027a. Vicente de Carvalho         112  95 102 97  92  113 105 74 
25  9º BPM  028a. Campinho                     61  60  68 89  80   49  48 62 
26  9º BPM  029a. Madureira                    64  67  89 59  59   76  54 40 
27  9º BPM  030a. Marechal Hermes              49  49  70 52  70   51  62 45 
28  41º BPM 031a. Ricardo Albuquerque         101 133 107 92  76   81  81 87 
29  18º BPM 032a. Taquara                      27  22  32 36  33   11  16 14 
30  14º BPM 033a. Realengo                    110  84 110 76  45   55  62 58 
⋮   ⋮       ⋮                                 ⋮   ⋮   ⋮   ⋮   ⋮   ⋮   ⋮   ⋮  
110 32º BPM 130a. Quissamã                     3  4    1   3  0   3   1   1  
111 25º BPM 132a. Arraial do Cabo              0  0    0   0  0   1   1   1  
112 8º BPM  134a. Campos                      19  6   11  10  4   4   8   4  
113 36º BPM 135a. Itaocara                     0  0    0   0  0   0   0   0  
114 36º BPM 136a. Santo Antonio de Pádua       0  0    0   1  0   0   0   0  
115 36º BPM 137a. Miracema                     0  0    0   0  0   0   0   0  
116 29º BPM 138a. Lajes de Muriaré             0  0    0   0  0   0   0   0  
117 29º BPM 139a. Porciúncula                  0  0    0   0  0   0   0   0  
118 29º BPM 140a. Natividade                   0  0    0   0  0   0   0   0  
119 8º BPM  141a. São Fidelis                  2  0    0   1  0   0   0   0  
120 36º BPM 142a. Cambucí                      1  0    0   0  0   0   0   0  
121 29º BPM 143a. Itaperuna                    0  0    0   1  0   0   0   0  
122 29º BPM 144a. Bom Jesus Itabapoana         0  0    0   0  0   0   0   0  
123 8º BPM  145a. São João da Barra            0  2    0   0  2   1   1   1  
124 8º BPM  146a. Guarus                       8  2    8   5  3   3   1   4  
125 8º BPM  147a. São Francisco de Itabapoana  5  1    1   1  1   0   1   1  
126 29º BPM 148a. Italva                       1  0    0   0  0   0   0   2  
127 11º BPM 151a. Nova Friburgo                1  0    0   0  0   1   2   0  
128 11º BPM 152a. Duas Barras                  0  0    0   0  0   0   0   0  
129 11º BPM 153a. Cantagal

## Lista de Exercícios 3

### Exercício 3.1

Coloque o conjunto de dados mtcars na área de trabalho;

In [56]:
df <- mtcars
df

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

### Exercício 3.2

Obter os nomes das variáiveis do conjunto de dados;



In [57]:
colnames(df)

[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[11] "carb"

### Exercício 3.3

Obter a quantidade de linhas e colunas do conjunto de dados;

In [58]:
sprintf("O número de linhas é equivalente à %i", nrow(df))
sprintf("O número de colunas é equivalente à %i", ncol(df))

[1] "O número de linhas é equivalente à 32"

[1] "O número de colunas é equivalente à 11"

### Exercício 3.4

Calcule a média da variável mpg;

In [59]:
mean(df[, 'mpg'])

[1] 20.09062

### Exercício 3.5

Obtenha os valores máximo e mínimo da variável disp;

In [60]:
sprintf("O valor máximo é equivalente à %.2f", max(df[, 'disp']))
sprintf("O valor máximo é equivalente à %.2f", min(df[, 'disp']))

[1] "O valor máximo é equivalente à 472.00"

[1] "O valor máximo é equivalente à 71.10"

### Exercício 3.6

Calcule a correlação entre as variáveis mpg e wt;

In [61]:
cor(df$mpg, df$wt)

[1] -0.8676594

### Exercício 3.7

Obtenha os registros para os quais a variável mpg é maior que 20 e hp é maior que 90;

In [62]:
df[which(df$mpg > 20 & df$hp > 90),]

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4      21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag  21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710     22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive 21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Merc 230       22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Toyota Corona  21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Porsche 914-2  26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa   30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Volvo 142E     21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

### Exercício 3.8

Substitua o nome da variável carb por qtd_carburadores;


In [63]:
colnames(df)[11] <- "qtd_carburadores"
colnames(df)

[1] "mpg"              "cyl"              "disp"             "hp"              
 [5] "drat"             "wt"               "qsec"             "vs"              
 [9] "am"               "gear"             "qtd_carburadores"

### Exercício 3.9

Remova a variável qsec do cojunto de dados;

In [64]:
within(df, rm('qsec'))

mpg  cyl disp  hp  drat wt    vs am gear qtd_carburadores
Mazda RX4           21.0 6   160.0 110 3.90 2.620 0  1  4    4               
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 0  1  4    4               
Datsun 710          22.8 4   108.0  93 3.85 2.320 1  1  4    1               
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 1  0  3    1               
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 0  0  3    2               
Valiant             18.1 6   225.0 105 2.76 3.460 1  0  3    1               
Duster 360          14.3 8   360.0 245 3.21 3.570 0  0  3    4               
Merc 240D           24.4 4   146.7  62 3.69 3.190 1  0  4    2               
Merc 230            22.8 4   140.8  95 3.92 3.150 1  0  4    2               
Merc 280            19.2 6   167.6 123 3.92 3.440 1  0  4    4               
Merc 280C           17.8 6   167.6 123 3.92 3.440 1  0  4    4               
Merc 450SE          16.4 8   275.8 180 3.07 4.070 0  0  3    3               
Merc 450SL          17.3 8   275.8 180 3.07 3.730 0  0  3    3               
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 0  0  3    3               
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 0  0  3    4               
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 0  0  3    4               
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 0  0  3    4               
Fiat 128            32.4 4    78.7  66 4.08 2.200 1  1  4    1               
Honda Civic         30.4 4    75.7  52 4.93 1.615 1  1  4    2               
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 1  1  4    1               
Toyota Corona       21.5 4   120.1  97 3.70 2.465 1  0  3    1               
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 0  0  3    2               
AMC Javelin         15.2 8   304.0 150 3.15 3.435 0  0  3    2               
Camaro Z28          13.3 8   350.0 245 3.73 3.840 0  0  3    4               
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 0  0  3    2               
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 1  1  4    1               
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 0  1  5    2               
Lotus Europa        30.4 4    95.1 113 3.77 1.513 1  1  5    2               
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 0  1  5    4               
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 0  1  5    6               
Maserati Bora       15.0 8   301.0 335 3.54 3.570 0  1  5    8               
Volvo 142E          21.4 4   121.0 109 4.11 2.780 1  1  4    2

### Exercício 3.10

Obtenha um novo data frame contendo apenas as variáveis mpg, wt e cyl;

In [65]:
new_df <- df[c("mpg", "wt", "cyl")]
new_df

mpg  wt    cyl
Mazda RX4           21.0 2.620 6  
Mazda RX4 Wag       21.0 2.875 6  
Datsun 710          22.8 2.320 4  
Hornet 4 Drive      21.4 3.215 6  
Hornet Sportabout   18.7 3.440 8  
Valiant             18.1 3.460 6  
Duster 360          14.3 3.570 8  
Merc 240D           24.4 3.190 4  
Merc 230            22.8 3.150 4  
Merc 280            19.2 3.440 6  
Merc 280C           17.8 3.440 6  
Merc 450SE          16.4 4.070 8  
Merc 450SL          17.3 3.730 8  
Merc 450SLC         15.2 3.780 8  
Cadillac Fleetwood  10.4 5.250 8  
Lincoln Continental 10.4 5.424 8  
Chrysler Imperial   14.7 5.345 8  
Fiat 128            32.4 2.200 4  
Honda Civic         30.4 1.615 4  
Toyota Corolla      33.9 1.835 4  
Toyota Corona       21.5 2.465 4  
Dodge Challenger    15.5 3.520 8  
AMC Javelin         15.2 3.435 8  
Camaro Z28          13.3 3.840 8  
Pontiac Firebird    19.2 3.845 8  
Fiat X1-9           27.3 1.935 4  
Porsche 914-2       26.0 2.140 4  
Lotus Europa        30.4 1.513 4  
Ford Pantera L      15.8 3.170 8  
Ferrari Dino        19.7 2.770 6  
Maserati Bora       15.0 3.570 8  
Volvo 142E          21.4 2.780 4

### Exercício 3.11

No data frame obtido na questão anterior reordene as colunas de forma que as variáveis apareçam na seguinte ordem: cyl, mpg, wt;

In [66]:
cols_ordered = order(names(new_df))
new_df[cols_ordered]

cyl mpg  wt   
Mazda RX4           6   21.0 2.620
Mazda RX4 Wag       6   21.0 2.875
Datsun 710          4   22.8 2.320
Hornet 4 Drive      6   21.4 3.215
Hornet Sportabout   8   18.7 3.440
Valiant             6   18.1 3.460
Duster 360          8   14.3 3.570
Merc 240D           4   24.4 3.190
Merc 230            4   22.8 3.150
Merc 280            6   19.2 3.440
Merc 280C           6   17.8 3.440
Merc 450SE          8   16.4 4.070
Merc 450SL          8   17.3 3.730
Merc 450SLC         8   15.2 3.780
Cadillac Fleetwood  8   10.4 5.250
Lincoln Continental 8   10.4 5.424
Chrysler Imperial   8   14.7 5.345
Fiat 128            4   32.4 2.200
Honda Civic         4   30.4 1.615
Toyota Corolla      4   33.9 1.835
Toyota Corona       4   21.5 2.465
Dodge Challenger    8   15.5 3.520
AMC Javelin         8   15.2 3.435
Camaro Z28          8   13.3 3.840
Pontiac Firebird    8   19.2 3.845
Fiat X1-9           4   27.3 1.935
Porsche 914-2       4   26.0 2.140
Lotus Europa        4   30.4 1.513
Ford Pantera L      8   15.8 3.170
Ferrari Dino        6   19.7 2.770
Maserati Bora       8   15.0 3.570
Volvo 142E          4   21.4 2.780

### Exercício 3.12

Obenha os nomes das linhas do conjunto de dados;

In [67]:
rownames(new_df)

[1] "Mazda RX4"           "Mazda RX4 Wag"       "Datsun 710"         
 [4] "Hornet 4 Drive"      "Hornet Sportabout"   "Valiant"            
 [7] "Duster 360"          "Merc 240D"           "Merc 230"           
[10] "Merc 280"            "Merc 280C"           "Merc 450SE"         
[13] "Merc 450SL"          "Merc 450SLC"         "Cadillac Fleetwood" 
[16] "Lincoln Continental" "Chrysler Imperial"   "Fiat 128"           
[19] "Honda Civic"         "Toyota Corolla"      "Toyota Corona"      
[22] "Dodge Challenger"    "AMC Javelin"         "Camaro Z28"         
[25] "Pontiac Firebird"    "Fiat X1-9"           "Porsche 914-2"      
[28] "Lotus Europa"        "Ford Pantera L"      "Ferrari Dino"       
[31] "Maserati Bora"       "Volvo 142E"

### Exercício 3.13

Qual o valor da variável hp na linha 15 da base de dados?;

In [68]:
df[15, "hp"]

[1] 205

### Exercício 3.14

Qual a média dos valores da variável mpg quando a variável cyl é 6?;

In [69]:
df_x = df[which(df$cyl > 6),]
mean(df_x[, "mpg"])

[1] 15.1

### Exercício 3.15

Qual é o produto do valor máximo de cyl com o valor mínimo de wt?;


In [70]:
wt_min_value = which.min(df$wt)
cyl_max_value = which.max(df$cyl)

wt_min_value * cyl_max_value

[1] 140

### Exercício 3.16

Obtenha as linhas 4, 10 e 19 e as colunas mpg, drat e vs do conjunto de dados;

In [71]:
selected_cols = c("mpg", "drat", "vs")
df[c(4, 10, 19), selected_cols]

mpg  drat vs
Hornet 4 Drive 21.4 3.08 1 
Merc 280       19.2 3.92 1 
Honda Civic    30.4 4.93 1

### Exercício 3.17

Obtenha um vetor que seja a soma da variável cyl com a variável qtd_carburadoes. Qual é a soma dos valores deste novo vetor?

In [72]:
new_df[, "cyl"] + df[, "qtd_carburadores"]

[1] 10 10  5  7 10  7 12  6  6 10 10 11 11 11 12 12 12  5  6  5  5 10 10 12 10
[26]  5  6  6 12 12 16  6